src/lib.rs

```diff
@@ -0,0 +1,189 @@
+        println!("test");
+        let args: Vec<_> = args
+            .iter()
+            .map(|s| s.to_string_lossy())
+            .collect();
+        let args: HashMap<&str, &str> = args
+            .iter()
+            .map(|s| {
+                let mut parts = s.splitn(2, '=');
+                (parts.next().unwrap(), parts.next().unwrap_or(""))
+            })
+            .collect();
+        let client_id: &str = match args.get("client_id") {
+            Some(client_id) => client_id,
+            None => return PamResultCode::PAM_AUTH_ERR,
+        };
+        let client_secret: &str = match args.get("client_secret") {
+            Some(client_secret) => client_secret,
+            None => return PamResultCode::PAM_AUTH_ERR,
+        };
+        let guild_id: &str = match args.get("guild") {
+            Some(guild) => guild,
+            None => return PamResultCode::PAM_AUTH_ERR,
+        };
+        let role: &str = match args.get("role") {
+            Some(role) => role,
+            None => return PamResultCode::PAM_AUTH_ERR,
+        };
+        let conv = match pamh.get_item::<Conv>() {
+            Ok(Some(conv)) => conv,
+            Ok(None) => {
+                unreachable!("No conv available");
+            }
+            Err(err) => {
+                println!("Couldn't get pam_conv");
+                return err;
+            }
+        };
+        let state: &&str = &"unknown";
+        let rhost = match pamh.get_item::<RHost>() {
+            Ok(Some(rhost)) => rhost,
+            Ok(None) => {
+                unreachable!("No rhost available");
+            }
+            Err(err) => {
+                println!("Couldn't get pam_rhost");
+                return err;
+            }
+        };
+        let port = match get_free_tcp() {
+            Some(port) => port,
+            None => 0,
+        };
+        let authorize_url = format!("{}/oauth2/authorize?response_type=code&client_id={}&scope=guilds.members.read&state=fart&redirect_uri=http://{}:{}", API_ENDPOINT, client_id, rhost.to_str().unwrap(), port);    
+        pam_try!(conv.send(PAM_TEXT_INFO, &authorize_url));
+        let rt = Runtime::new().unwrap();
+        let gang = rt.block_on(generate_new_token(state, port));
+        let access_token = rt.block_on(get_access_token(client_id, client_secret, gang, format!("http://127.0.0.1:{}",port).to_string())).unwrap();
+        let actual_token_json = json::parse(&access_token).unwrap();
+        let actual_token = actual_token_json["access_token"].as_str().unwrap();
+        let member = rt.block_on(get_guild_member(actual_token, guild_id)).unwrap();
+        let member_json = json::parse(&member).unwrap();
+        if member_json["roles"].contains(role){
+            return PamResultCode::PAM_SUCCESS
+        }
+        pam_try!(conv.send(PAM_ERROR_MSG, "You do not have the required role!"));
+        PamResultCode::PAM_PERM_DENIED
+    }
```

Comment:
`sm_authenticate` gets pretty crammed to look at. This might be a taste thing for me, but I like having a line or two of whitespace between chunks of relevant code.